### PR TITLE
[SYCL] Fix hang in conditional variable pattern in program manager

### PR DESCRIPTION
**Pull request by alexanderfle**
The issue rarely occurs sporadically when we try to get a kernel instance by multiple threads:
`auto kernel = prog.get_kernel<some_kernel>();`

When debugging, all threads have finished except two threads, the main thread which is waiting for the last one to complete.
This last thread sleeps on condition_variable on the line:
`Cache.waitUntilBuilt() in sycl\source\detail\program_manager\program_manager.cpp`
It looks like the build has already been done, all other threads have already done their work and finished,  except for one thread that is waiting for notification in order to wake up.

Checked the condition variables pattern and found that `BuildResult->State` must be modified under the mutex.
Quote from doc:
> "Even if the shared variable is atomic, it must be modified under the mutex in order to correctly publish the modification to the waiting thread."

Signed-off-by: Alexander Flegontov <alexander.flegontov@intel.com>

### CODE REVIEW

**Review by s-kanaev — APPROVED** [review]
It's a shame we don't have test for this use-case.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -195,7 +195,12 @@ getOrBuild(KernelProgramCache &KPCache, KeyT &&CacheKey, AcquireFT &&Acquire,
     BuildResult->Ptr.store(Desired);
 #endif
 
-    BuildResult->State.store(BS_Done);
+    {
+      // Even if shared variable is atomic,it must be modified under the mutex
```
Comment (dm-vodopyanov) [inline sycl/source/detail/program_manager/program_manager.cpp:199]: Minor: ```atomic,it``` -> ```atomic, it```

**Review by dm-vodopyanov — APPROVED** [review]

**Review by kbobrovs — CHANGES_REQUESTED** [review]
We've got quite a few issues related to cache management. But this complex piece still does not have a design doc. Can we please create a section on the caches in the main design doc or a separate document?

**Comment by s-kanaev** [comment]
> Can we please create a section on the caches in the main design doc or a separate document?

Sorry, I forgot about this one. I will prepare the document.

**Review by dm-vodopyanov — APPROVED** [review]

**Comment by s-kanaev** [comment]
> > Can we please create a section on the caches in the main design doc or a separate document?

> Sorry, I forgot about this one. I will prepare the document.

I mean, in distinct patch.

**Comment by alexanderfle** [comment]
@kbobrovs , Since the design document will be added to a separate PR, could  you please review and approve the current PR?

**Review by kbobrovs — APPROVED** [review]

**Pull request merged**